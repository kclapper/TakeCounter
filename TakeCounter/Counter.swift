//
//  Counter.swift
//  TakeCounter
//
//  Created by Kyle Clapper on 1/15/24.
//

import Foundation

typealias Count = UInt
typealias CountEvent = Count

protocol Listener: AnyObject {
    func update(_ event: CountEvent) -> Void
}

protocol Counter {
    var count: Count { get set }
    
    func increment() -> Void
    func decrement() -> Void
    func reset() -> Void
}

protocol Emitter {
    func addListener(_ listener: Listener) -> Void
    func removeListener(_ listener: Listener) -> Void
    func updateListeners(_ event: CountEvent) -> Void
}

protocol Store {
    func saveCount(_ count: Count) -> Void
}

class SingletonCounter: Counter, Emitter {
    private static var instance: SingletonCounter? = nil
    public static func getInstance() -> SingletonCounter {
        if SingletonCounter.instance == nil {
            SingletonCounter.instance = SingletonCounter()
        }
        return SingletonCounter.instance!
    }
    
    var count: UInt
    var listeners: [any Listener]
    var dataStore: Store?
    
    private init() {
        self.count = 0
        self.listeners = []
    }
    
    func setDataStore(dataStore: Store) {
        self.dataStore = dataStore
    }
    
    func increment() -> Void {
        count += 1
        updateListeners(count)
        
        dataStore?.saveCount(count)
    }
    func decrement() -> Void {
        count -= 1
        updateListeners(count)
        
        dataStore?.saveCount(count)
    }
    func reset() -> Void {
        count = 0
        updateListeners(count)
        dataStore?.saveCount(count)
    }
    
    func addListener(_ listener: any Listener) {
        listeners.append(listener)
    }
    func removeListener(_ listener: any Listener) {
        listeners.removeAll(where: { $0 === listener })
    }
    func updateListeners(_ event: CountEvent) {
        listeners.forEach({ $0.update(event) })
    }
}
