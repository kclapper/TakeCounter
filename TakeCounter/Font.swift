//
//  Font.swift
//  TakeCounter
//
//  Created by Kyle Clapper on 2/18/24.
//

import SwiftUI

enum CustomFontWeight: String {
        case bold = "Avenir-Black"
//        case demibold = "Avenir-Black"
        case medium = "Avenir-Medium"
        case light = "Avenir-Light"
        case heavy = "Avenir-Heavy"
        case regular = "Avenir-Book"
//        case black = "Avenir-Black"
}

enum CustomFontStyle {
    case extraLargeTitle2
    case extraLargeTitle
    case largeTitle
    case title
    case title2
    case title3
    case headline
    case subheadline
    case body
    case callout
    case caption
    case caption2
    case footnote
}

struct CustomFontModifier: ViewModifier {
    @AppStorage("fontSize") private var fontSize: Double = 10
    
    let relativeTo: Font.TextStyle  = .body
    
    let name: CustomFontWeight
    let size: CGFloat
    
    init(_ name: CustomFontWeight, size: CGFloat) {
        self.name = name
        self.size = size
    }
    
    func body(content: Content) -> some View {
        content
            .font(.custom(name.rawValue, size: size + CGFloat(fontSize), relativeTo: relativeTo))
    }
}

extension View {
    func customFont(_ name: CustomFontWeight, size: CGFloat) -> some View {
        modifier(CustomFontModifier(name, size: size))
    }
    
    func customFont() -> some View {
        modifier(CustomFontModifier(.regular, size: CGFloat(10)))
    }
    
    func customFont(size: CGFloat) -> some View {
        modifier(CustomFontModifier(.regular, size: size))
    }
}
