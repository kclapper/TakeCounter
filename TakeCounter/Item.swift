//
//  Item.swift
//  TakeCounter
//
//  Created by Kyle Clapper on 1/15/24.
//

import Foundation
import SwiftData

@Model
final class Item {
    var timestamp: Date
    
    init(timestamp: Date) {
        self.timestamp = timestamp
    }
}
