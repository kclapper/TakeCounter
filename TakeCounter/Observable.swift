//
//  Observable.swift
//  TakeCounter
//
//  Created by Kyle Clapper on 1/15/24.
//

import Foundation

//protocol Listener: Equatable {
//    associatedtype Event
//    func update(_ event: Event) -> Void
//}
//
//protocol Observable {
//    associatedtype Event
//    associatedtype EventListener: Listener where EventListener.Event == Event
//    
//    var listeners: [EventListener] { get set }
//}
//
//extension Observable {
//    mutating func addListener(listener: EventListener) -> Void {
//        listeners.append(listener)
//    }
//    mutating func removeListener(listener: EventListener) -> Void {
//        listeners.removeAll(where: { $0 == listener })
//    }
//    func updateListeners(event: Event) -> Void {
//        for listener in listeners {
//            listener.update(event)
//        }
//    }
//}
