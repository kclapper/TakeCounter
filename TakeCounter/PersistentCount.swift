//
//  PersistentCount.swift
//  TakeCounter
//
//  Created by Kyle Clapper on 1/15/24.
//

import Foundation
import SwiftData

@Model
final class PersistentCount: Store {
    var count: Count
    
    init(count: Count) {
        self.count = count
    }
    
    func saveCount(_ count: Count) {
        self.count = count
    }
}
