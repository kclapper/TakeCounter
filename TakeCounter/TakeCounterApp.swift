//
//  TakeCounterApp.swift
//  TakeCounter
//
//  Created by Kyle Clapper on 1/15/24.
//

import SwiftUI
import SwiftData
import AppKit

let defaultWidth = CGFloat(550)
let defaultHeight = CGFloat(250)

@main
struct TakeCounterApp: App {
    
    var body: some Scene {
        WindowGroup {
            VStack {
                HStack {
                    ClientDetailView()
                    Spacer()
                }
                Spacer()
                CounterView()
            }
            .padding()
            .background(.background)
            .removeFocusOnTap()
        }
        .defaultSize(width: defaultWidth, height: defaultHeight)
    }
}

#Preview {
    Group {
        VStack {
            HStack {
                ClientDetailView()
                Spacer()
            }
            Spacer()
            CounterView()
        }
            .frame(width: defaultWidth, height: defaultHeight)
            .background(.background)
            .removeFocusOnTap()
    }
}
