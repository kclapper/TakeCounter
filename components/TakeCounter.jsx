import React from 'react';

import './index.scss';

import { SettingsProvider } from './Settings';
import SettingsMenu from './Settings/Menu';

import Details from './Details';
import Counter from './Counter';

export default function TakeCounter() {
  return <SettingsProvider>
           <div className="d-flex flex-column text-white" style={{ minHeight: "100vh" }}>
             <div className="d-flex justify-content-between ms-4 me-4 mt-2 mb-0 m-sm-4">
               <Details />
               <SettingsMenu />
             </div>
             <Counter />
           </div>
         </SettingsProvider>
}
