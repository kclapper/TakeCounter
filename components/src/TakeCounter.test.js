import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { expect, it } from '@jest/globals';

import TakeCounter from './TakeCounter';

it('Matches snapshot', async () => {
  const { asFragment } = render(<TakeCounter />);
  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});
