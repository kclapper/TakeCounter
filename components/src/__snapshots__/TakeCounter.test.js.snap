// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column text-white"
    style="min-height: 100vh;"
  >
    <div
      class="d-flex justify-content-between ms-4 me-4 mt-2 mb-0 m-sm-4"
    >
      <div
        class="d-flex justify-content-start align-items-stretch"
      >
        <div
          class="d-none d-sm-block"
          style="flex-grow: 0.25;"
        />
        <div
          class="h3 border-bottom m-2 ms-0"
          contenteditable="true"
          style="outline: none; min-width: 25vw;"
        />
        <div
          class="d-none d-sm-block"
          style="flex-grow: 2;"
        />
      </div>
      <div>
        <button
          aria-controls="settingsMenu"
          class="btn"
          data-bs-target="#settingsMenu"
          data-bs-toggle="offcanvas"
          type="button"
        >
          <svg
            class="bi bi-gear-wide-connected"
            fill="white"
            height="24"
            viewBox="0 0 16 16"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.068.727c.243-.97 1.62-.97 1.864 0l.071.286a.96.96 0 0 0 1.622.434l.205-.211c.695-.719 1.888-.03 1.613.931l-.08.284a.96.96 0 0 0 1.187 1.187l.283-.081c.96-.275 1.65.918.931 1.613l-.211.205a.96.96 0 0 0 .434 1.622l.286.071c.97.243.97 1.62 0 1.864l-.286.071a.96.96 0 0 0-.434 1.622l.211.205c.719.695.03 1.888-.931 1.613l-.284-.08a.96.96 0 0 0-1.187 1.187l.081.283c.275.96-.918 1.65-1.613.931l-.205-.211a.96.96 0 0 0-1.622.434l-.071.286c-.243.97-1.62.97-1.864 0l-.071-.286a.96.96 0 0 0-1.622-.434l-.205.211c-.695.719-1.888.03-1.613-.931l.08-.284a.96.96 0 0 0-1.186-1.187l-.284.081c-.96.275-1.65-.918-.931-1.613l.211-.205a.96.96 0 0 0-.434-1.622l-.286-.071c-.97-.243-.97-1.62 0-1.864l.286-.071a.96.96 0 0 0 .434-1.622l-.211-.205c-.719-.695-.03-1.888.931-1.613l.284.08a.96.96 0 0 0 1.187-1.186l-.081-.284c-.275-.96.918-1.65 1.613-.931l.205.211a.96.96 0 0 0 1.622-.434zM12.973 8.5H8.25l-2.834 3.779A4.998 4.998 0 0 0 12.973 8.5m0-1a4.998 4.998 0 0 0-7.557-3.779l2.834 3.78zM5.048 3.967l-.087.065zm-.431.355A4.98 4.98 0 0 0 3.002 8c0 1.455.622 2.765 1.615 3.678L7.375 8zm.344 7.646.087.065z"
            />
          </svg>
        </button>
        <div
          aria-labelledby="settingsMenu"
          class="w-100 offcanvas offcanvas-start overflow-y-scroll p-4"
          data-bs-theme="dark"
          id="settingsMenu"
          tabindex="-1"
        >
          <div
            class="offcanvas-header"
          >
            <button
              aria-label="Close"
              class="btn-close"
              data-bs-dismiss="offcanvas"
              type="button"
            />
          </div>
          <div
            class="container"
          >
            <h4
              class="row border-bottom"
            >
              Take Counting
            </h4>
            <div
              class="row mb-2"
            >
              <div
                class="col-6 align-self-center p-0"
              >
                <h5
                  class="my-0"
                >
                  Mode
                </h5>
              </div>
              <div
                class="col-6 align-self-center p-0"
              >
                <select
                  aria-label="Mode"
                  class="form-select"
                >
                  <option
                    value="manual"
                  >
                    Manual
                  </option>
                  <option
                    value="ptFileWatcher"
                  >
                    Pro Tools File Watcher
                  </option>
                </select>
              </div>
            </div>
            <div>
              <h4
                class="row border-bottom mt-3"
              >
                Keyboard Shortcuts
              </h4>
              <p>
                To change keyboard shortcuts: click the shortcut box, enter the shortcut you want, then click away from the box.
              </p>
              <div
                class="row mb-2"
              >
                <div
                  class="col-6 align-self-center p-0"
                >
                  <h5
                    class="my-0"
                  >
                    Increment count
                  </h5>
                </div>
                <div
                  class="col-6 align-self-center p-0"
                >
                  <div
                    class="border rounded p-1"
                    contenteditable="true"
                    style="outline: none; min-width: 20%;"
                  />
                </div>
              </div>
              <div
                class="row mb-2"
              >
                <div
                  class="col-6 align-self-center p-0"
                >
                  <h5
                    class="my-0"
                  >
                    Decrement count
                  </h5>
                </div>
                <div
                  class="col-6 align-self-center p-0"
                >
                  <div
                    class="border rounded p-1"
                    contenteditable="true"
                    style="outline: none; min-width: 20%;"
                  />
                </div>
              </div>
              <div
                class="row mb-2"
              >
                <div
                  class="col-6 align-self-center p-0"
                >
                  <h5
                    class="my-0"
                  >
                    Reset count
                  </h5>
                </div>
                <div
                  class="col-6 align-self-center p-0"
                >
                  <div
                    class="border rounded p-1"
                    contenteditable="true"
                    style="outline: none; min-width: 20%;"
                  />
                </div>
              </div>
            </div>
            <div
              class="row justify-content-start border-top mt-3 pt-2"
            >
              <div
                class="row mb-2"
              >
                <div
                  class="col-6 align-self-center p-0"
                >
                  <h5
                    class="my-0"
                  >
                    Reset
                  </h5>
                </div>
                <div
                  class="col-6 align-self-center p-0"
                >
                  <div
                    class="d-flex"
                  >
                    <div
                      class="flex-grow-1"
                    />
                    <button
                      class="btn btn-outline-secondary"
                      data-bs-placement="top"
                      data-bs-toggle="tooltip"
                      type="button"
                    >
                      Reset Default Settings
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="d-flex flex-column align-items-center justify-content-center"
      style="flex-grow: 1;"
    >
      <div
        style="flex-grow: 0.75;"
      />
      <div
        class="d-flex"
      >
        <h1
          class="display-1 me-4 me-md-5"
        >
          Take
        </h1>
        <div
          class="display-1"
          contenteditable="true"
          style="outline: none; min-width: 40%;"
        />
      </div>
      <div
        class=""
        style="flex-grow: 0;"
      >
        <button
          class="btn btn-primary m-1"
          data-bs-placement="bottom"
          data-bs-title="Alt+Shift+-"
          data-bs-toggle="tooltip"
          type="button"
        >
          -
        </button>
        <button
          class="btn btn-primary m-1"
          data-bs-placement="bottom"
          data-bs-toggle="tooltip"
          type="button"
        >
          reset
        </button>
        <button
          class="btn btn-primary m-1"
          data-bs-placement="bottom"
          data-bs-title="Alt+Shift+="
          data-bs-toggle="tooltip"
          type="button"
        >
          +
        </button>
      </div>
      <div
        style="flex-grow: 1;"
      />
    </div>
  </div>
</DocumentFragment>
`;
